Add benchmark runners user guide (#55)
aai-institute/nnbench: docs/guides/runners.md

# Running benchmarks with the benchmark runner

> The `BecnhmarkRunner` is `nnbench`'s abstract interface to collect and execute benchmarks.

Reviewer: ```suggestion
nnbench provides the `BenchmarkRunner` as a compact interface to collect and run benchmarks selectively.
```

> You can either use it directly or instantiate a derivative class to override selected methods and implement you own logic.

Reviewer: I would drop this, since it is a no-content sentence (this is true for almost every Python type).

##  The abstract `BenchmarkRunner`  class
Let's first instantiate and then walk through the abstract class.

```python
from nnbench import BenchmarkRunner

runner = BenchmarkRunner()
```

> Now we can use the `BenchmarkRunner.collect` method to collect benchmarks from files or directories.
> Lets assume we have the following benchmark setup:

Reviewer: ```suggestion
Use the `BenchmarkRunner.collect` method to collect benchmarks from files or directories.
Assume we have the following benchmark setup:

```

import nnbench
import numpy as np
import time
from sklearn import base
from memory_profiler import memory_usage

@nnbench.benchmark(tags=("inference-resources",))
def avg_inference_time(model: base.BaseEstimator, X_test: np.ndarray, steps: int) -> float:
    total_time = 0.0

    for _ in range(steps):
        start_time = time.time()
        model.predict(X_test)
        end_time = time.time()
        total_time += end_time - start_time

    average_time = total_time / steps
    return average_time

@nnbench.benchmark(tags=("inference-resources",))
def memory_usage_during_inference(model: base.BaseEstimator, X_test: np.ndarray) -> float:
    def predict():
        model.predict(X_test)

    mem_usage = memory_usage(predict, max_usage=True)
    return mem_usage
```

```python
# ./model_info/storage_resources.py
import nnbench
import numpy as np
from sklearn import base

@nnbench.benchmark
def model_size_mb(model: base.BaseEstimator) -> float:
    total_size_bytes = 0

    for attr, value in model.__dict__.items():
        if isinstance(value, np.ndarray):
            total_size_bytes += value.nbytes
        elif isinstance(value, list) or isinstance(value, dict):

>             for item in value:

Reviewer: For a `dict` value, this does not do what you think it does: `dict.__iter__` gives an iterator over the keys, not the values (which you are looking for here).

                if isinstance(item, np.ndarray):
                    total_size_bytes += item.nbytes

    size_mb = total_size_bytes / (1024 * 1024)

>     return size_mb

Reviewer: This gives the size in MiB, not MB (for that, you need to divide by 1e6).

```

Now we can collect benchmarks from files:

```python

> runner.collect('./performance/model_benchmark.py')

Reviewer: ```suggestion
runner.collect('performance/model_benchmark.py')
```

Similarly for the filenames in comments above the code examples earlier.

```
Or directories:

```python
runner.collect('./model_info')
```

This collection can happen iteratively. So, after executing the two collections our runner has all four benchmarks ready for execution.

> If we want to remove the collected benchmarks, we can use the `BenchmarkRunner.clear` method.

Reviewer: ```suggestion
To remove the collected benchmarks again, use the `BenchmarkRunner.clear()` method.
```

> We can also supply tags to the collection. Then the runner will only collect benchmarks with the appropriate tag.

Reviewer: ```suggestion
You can also supply tags to the runner to selectively collect only benchmarks with the appropriate tag.
```

> Hence, we can - after clearing the runner - collect all benchmarks with the "inference-resources" tag:

Reviewer: ```suggestion
For example, after clearing the runner again, you can collect all benchmarks with the "inference-resources" tag as such:
```

```python
runner.collect('./model_info', tags=("inference-resources",))
```
Note the trailing comma which is necessary for single entires in parenthesis to be interpreted as a tuple, the proper datatypes for tags.

> We can execute our benchmark with the `BenchmarkRunner.run` method. However, we have to pass all the necessary parameters for our benchmark as well. 

Reviewer: ```suggestion
To run the benchmarks, call the `BenchmarkRunner.run()` method and supply the necessary parameters required by the collected benchmarks.
```

If we can import a scikit-learn `model` from another file, let's say `traininf.py`, the code to execute the "inference-resources" benchmarks looks like this: 

# TODO: currently this does not work. We have not added a path to the runner. Should it work without?
```python
from training import model, X_test

runner.run({"model": model, "X_test": X_test, "steps": 25})
```

> ## Custom benchmark runners

Reviewer: Remove this and call it a day - we can amend the guide again after there's actual need to implement a custom runner.

> # Running benchmarks with the benchmark runner

Reviewer: ```suggestion
# Collecting and running benchmarks
```

nnbench provides the `BenchmarkRunner` as a compact interface to collect and run benchmarks selectively.

##  The abstract `BenchmarkRunner`  class

> Let's first instantiate and then walk through the abstract class.

Reviewer: Maybe "base" instead of abstract? Abstract reminds me of `abc`, which we explicitly do not use.

```python
from nnbench import BenchmarkRunner

runner = BenchmarkRunner()
```

> Use the `BenchmarkRunner.collect` method to collect benchmarks from files or directories.  

Reviewer: ```suggestion
Use the `BenchmarkRunner.collect()` method to collect benchmarks from files or directories.  
```

Assume we have the following benchmark setup:
```python
# dir_a/bm1.py
import nnbench

@nnbench.benchmark
def dummy_benchmark(a: int) -> int:
    return a
```

```python
# dir_b/bm2.py
import nnbench

@nnbench.benchmark(tags=("tag",))
def another_benchmark(b: int) -> int:
    return b

@nnbench.benchmark
def yet_another_benchmark(c: int) -> int:
    return c
```

```python
# dir_b/bm3.py
import nnbench
@nnbench.benchmark(tags=("tag",))
def the_last_benchmark(d: int) -> int:
    return d
```

Now we can collect benchmarks from files:

```python
runner.collect('dir_a/bm1.py')
```
Or directories:

```python
runner.collect('dir_b')
```

This collection can happen iteratively. So, after executing the two collections our runner has all four benchmarks ready for execution.

To remove the collected benchmarks again, use the `BenchmarkRunner.clear()` method.
You can also supply tags to the runner to selectively collect only benchmarks with the appropriate tag.

> For example, after clearing the runner again, you can collect all benchmarks with the "tag" tag as such:

Reviewer: ```suggestion
For example, after clearing the runner again, you can collect all benchmarks with the `"tag"` tag as such:
```

```python
runner.collect('dir_b', tags=("tag",))
```

> Note the trailing comma which is necessary for single entires in parenthesis to be interpreted as a tuple, the proper datatypes for tags.

Reviewer: I would leave this out, since modern IDEs catch this.
